File: 03_patterns/swift/mvvm_clean_arc/Clean/Data/Repository/TaskRepositoryImpl.swift
import Foundation

final class TaskRepositoryImpl: TaskRepository {
    private let localDataSource: TaskLocalDataSource

    init(localDataSource: TaskLocalDataSource) {
        self.localDataSource = localDataSource
    }

    func getTasks() -> [Task] {
        localDataSource.fetchTasks()
    }

    func createTask(title: String) -> [Task] {
        localDataSource.saveTask(title: title)
    }

    func deleteTask(id: UUID) -> [Task] {
        localDataSource.removeTask(id: id)
    }
}

File: 03_patterns/flutter/lib/clean/data/repository/task_repository_impl.dart
import '../../domain/entities/task.dart';
import '../../domain/repositories/task_repository.dart';
import '../data_source/task_local_data_source.dart';

class TaskRepositoryImpl implements TaskRepository {
  TaskRepositoryImpl({required TaskLocalDataSource localDataSource})
      : _localDataSource = localDataSource;

  final TaskLocalDataSource _localDataSource;

  @override
  List<Task> getTasks() => _localDataSource.fetchTasks();

  @override
  List<Task> createTask({required String title}) {
    return _localDataSource.saveTask(title: title);
  }

  @override
  List<Task> deleteTask({required String id}) {
    return _localDataSource.removeTask(id: id);
  }
}
